### DISCUSSION TITLE
Custom server with rsbuild dev server

### DISCUSSION THREAD

**Discussion by sowtame** [web-infra-dev/rsbuild#2046] [Q&A] [answered]
guys, if there is an example of using a custom server that would render an application on express, for example, using hmr and all the optimizations that are in rspack and rsbuild, and not just restart the server for every file change

a little demo of how I'm trying to use it https://github.com/sowtame/rsbuild-basic/blob/master/rsbuild.config.ts

**Comment by 9aoy** [comment]
Here's a guide on integrating the Rsbuild dev server with Express, which may be helpful for you.
https://rsbuild.dev/api/javascript-api/instance#rsbuildcreatedevserver

**Reply by chenjiahan** [reply]
@9aoy can we add an example to https://github.com/rspack-contrib/rspack-examples?

**Reply by 9aoy** [reply, marked as answer]
@chenjiahan  https://github.com/rspack-contrib/rspack-examples/pull/50  cc~

**Reply by sowtame** [reply]
thank you very much!! it helped to integrate the dev server with my server on express, but there is still a problem how to combine the render function, which located in the folder "src/server/server-render.tsx" (or the whole express app located in "src/server/index.tsx") with the dev server, so that there is hmr, optimization of rebuilding the js file and everything that rsbuild is very good at)
https://github.com/sowtame/rsbuild-basic/blob/master/rsbuild/devServer/server.mjs#L15

I want to create a small application with ssr in prod and dev builds (also created custom rsbuild assets plugins for it)

@9aoy @chenjiahan

**Reply by 9aoy** [reply]
@sowtame Get your question. Rsbuild currently does not provide an example of SSR.  Let me think about this later. 🤔

**Reply by 9aoy** [reply]
@sowtame  I wrote an example about using ssr with rsbuild, you can refer to this: https://github.com/9aoy/rsbuild-ssr-example/tree/main.
And we plan to provide some APIs in rsbuild for better use of ssr.  https://github.com/web-infra-dev/rsbuild/discussions/2221

**Reply by sowtame** [reply]
Oh, thank you so much for your time! @9aoy 

I will really look forward to improvements for convenient work with ssr, looking at other builders it seems to be a very cool and unique feature for rsbuild!

**Reply by sowtame** [reply]
also is it possible to enable restart whole server? for example when you add new api endpoint for some dev things...

Restart the server when the server side changes, but when developing the client side, it works the same as in your example, for example

**Reply by 9aoy** [reply]
About 'restart server', when you use a custom server, it is not started by rsbuild-cli, so it is not controlled by rsbuild.
But I know some node.js tools can support this. For example, [nodemon](https://www.npmjs.com/package/nodemon), you can change the startup script from `node ./rsbuild/devServer/index.mjs` to `nodemon --watch rsbuild ./rsbuild/devServer/index.mjs`, so that when some files in the `rsbuild/` directory change , the application will automatically restart.

**Reply by sowtame** [reply]
thanks, I was able to solve this using the msbuild api
https://github.com/sowtame/rsbuild-server-client-side/blob/main/server.mjs#L86

**Reply by sowtame** [reply]
I found a small bug when running your example
https://github.com/9aoy/rsbuild-ssr-example/tree/main

1 pnpm dev:ssr
2 change something in src/index.server.tsx (add console log for example)
3 open 2 tabs with http://localhost:3000/ and you can see a looped server call

**Reply by 9aoy** [reply]
@sowtame Reproduced your problem, let me find out what the problem is!

**Reply by omarkhatibgg** [reply]
Hey @9aoy , 
I just tried the example, on SSR it only serves the HTML from the index file, it does not render the react application till it hit the client.
update: I just noticed the `dev:ssr` command, but anyway, `rsbuild preivew` it seems it's not serving the server, I think it's using the index.html file

**Reply by 9aoy** [reply]
Hi, I rewrote the ssr example based on the environment API. Now you can refer to these two examples:
- ssr examples:  [ssr with devMiddleware](https://github.com/rspack-contrib/rspack-examples/tree/main/rsbuild/ssr)、[ssr with custom server](https://github.com/rspack-contrib/rspack-examples/tree/main/rsbuild/ssr-express)
- related PR: https://github.com/web-infra-dev/rsbuild/pull/2528

And the preview command is only used for local preview, it does not support powerful server capabilities.

**Reply by 9aoy** [reply]
@omarkhatibco  I wrote an example about the use of ssr + express in production mode, which may be helpful to you~
 https://github.com/rspack-contrib/rspack-examples/blob/main/rsbuild/ssr-express/prod-server.mjs

**Reply by omarkhatibgg** [reply]
@9aoy thank you! Will check it today
